lib/pos_epoch_complete_hook.go

```diff
@@ -45,62 +65,168 @@ func (bav *UtxoView) RunEpochCompleteHook(blockHeight uint64) error {
 }
+
+func (bav *UtxoView) generateAndSnapshotValidatorSet(epochNumber uint64) ([]*ValidatorEntry, error) {
+	// Snapshot the current top n active validators as the validator set.
+	validatorSet, err := bav.GetTopActiveValidatorsByStakeAmount(
+		bav.GetCurrentGlobalParamsEntry().ValidatorSetMaxNumValidators,
+	)
+	if err != nil {
+		return nil, errors.Wrapf(err, "generateAndSnapshotValidatorSet: error retrieving top ValidatorEntries: ")
+	}
+	for _, validatorEntry := range validatorSet {
+		bav._setSnapshotValidatorSetEntry(validatorEntry, epochNumber)
+	}
+
+	// Snapshot the current validator set's total stake. Note, the validator set is already filtered to the top n
+	// active validators for the epoch. The total stake is the sum of all of the active validators' stakes.
+	validatorSetTotalStakeAmountNanos := SumValidatorEntriesTotalStakeAmountNanos(validatorSet)
+	bav._setSnapshotValidatorSetTotalStakeAmountNanos(validatorSetTotalStakeAmountNanos, epochNumber)
+
+	return validatorSet, nil
+}
+
+func (bav *UtxoView) generateAndSnapshotLeaderSchedule(epochNumber uint64) error {
+	// Generate a random leader schedule and snapshot it.
+	leaderSchedule, err := bav.GenerateLeaderSchedule()
+	if err != nil {
+		return errors.Wrapf(err, "generateAndSnapshotLeaderSchedule: problem generating leader schedule: ")
+	}
+
+	for index, validatorPKID := range leaderSchedule {
+		if index > math.MaxUint16 {
+			return errors.Errorf("generateAndSnapshotLeaderSchedule: LeaderIndex %d overflows uint16", index)
+		}
+		bav._setSnapshotLeaderScheduleValidator(validatorPKID, uint16(index), epochNumber)
+	}
+
+	return nil
+}
+
+func (bav *UtxoView) generateAndSnapshotStakesToReward(epochNumber uint64, validatorSet []*ValidatorEntry) error {
+	// Fetch the current top n stake entries. Note, this query will return the top n stake entries regardless of whether
+	// or not they are in the active validator set of the epoch. By not filtering the stake entries while seeking through
+	// the DB, we guarantee that we will at most seek through n stake entries. This avoids the edge cases where the global
+	// set of stake entries is very large, and as a result of additional filters in the seek, we end up seeking through
+	// the entire global set.
+	topStakeEntries, err := bav.GetTopStakesByStakeAmount(bav.GetCurrentGlobalParamsEntry().StakingRewardsMaxNumStakes)
+	if err != nil {
+		return errors.Wrapf(err, "RunEpochCompleteHook: error retrieving top StakeEntries: ")
+	}
+
+	// Filter the top n stake entries by the current validator set. We do not want to reward stakes that are
+	// not in the current validator set, so we pre-filter them here before snapshotting them.
+	validatorSetPKIDs := NewSet([]PKID{})
+	for _, validatorEntry := range validatorSet {
+		validatorSetPKIDs.Add(*validatorEntry.ValidatorPKID)
+	}
+	topStakesInValidatorSet := collections.SliceFilter(topStakeEntries, func(s *StakeEntry) bool {
+		return validatorSetPKIDs.Includes(*s.ValidatorPKID)
+	})
```

Comment:
So just wanted to clarify: I think what this will do is we will get back the max stake entries, say we get back 10k, and then cull out stake entries that are attached to either jailed validators or validators who don't fit in the top validator set.

I think this makes sense. We should definitely cut out the jailed validators. I'm just kinda wondering if we should still reward people who stake to validators who don't participate in consensus (i.e. who are beyond the max validators). The only reason I can see for incentivizing that is to prevent a situation where people only want to stake to the top ten validators or something, for fear of not being rewarded. For example, this rule would certainly hurt the Nth validator, and probably up to the N-100th or so.

Anyway I think it's fine. Let's leave it as is. Just lmk if I misunderstood anything above.

Reply:
@diamondhands0 you've understood correctly. The fact that we additionally filter stakers after fetching the top 10k has the downside you mentioned.

An alternative approach I was considering was to make the DB seek over the top stake entries dynamic. I.e. while doing the seek, we can filter out stake entries that are not part of the top n validators as well. This way, we're guaranteed to get 10k stake entries and they are ALL part of the active validator set. 

The downside of the alternative approach is that, it results in an uncapped DB seek. Meaning, if we have 10 million stake entries, but we want to fetch the top 10k that are part of the active validator set, we may end up seeking very deep into the list of 10 mil stake entries that satisfy the active validator set filter. This could be costly.

Reply:
I was messaging in Slack but going to consolidate my thoughts here.

I think that leaving it as-is is fine for now. Once we hit the 10k stake threshold, this is easy enough to fix with a hard fork.

The above being said, if we wanted to do this the "right" way, then I think it would actually involve the following:
1. Include the validator status in `PrefixStakeByStakeAmount` at the beginning like we do for validators themselves. This will make it so that the 10k stakers we fetch are attached to unjailed validators.
2. Then, if you want, you can eliminate the check on validators being in the top N. This would resolve the perverse incentive to avoid smaller validators, since you get paid as long as your boy isn't jailed.

Reply:
Nevermind. I realized you can't limit to the validators in the top N because then you wouldn't be able to figure out their commissions. I think every chain has this issue though so it's nbd.

I'd just leave this as is. Once we hit the 10k threshold we can apply whatever the proper fix is.

Reply:
Updated up here to consolidate our chat on Slack too. The solution I proposed above expresses quite simply in the code actually. You can check it out here: https://github.com/deso-protocol/core/pull/632

After seeing this solution in action, I'd be pretty happy to just ship it now and have one less nuance to think about in the top 10k filtering for staker snapshots.